src/Workspaces/Core/Portable/Shared/Utilities/RoslynParallel.NetFramework.cs

```diff
@@ -0,0 +1,655 @@
+            /// <summary>
+            /// The delegate to invoke on each worker to run the enumerator processing loop.
+            /// </summary>
+            /// <remarks>
+            /// This could have been an action rather than a func, but it returns a task so that the task body is an async Task
+            /// method rather than async void, even though the worker body catches all exceptions and the returned Task is ignored.
+            /// </remarks>
+            private readonly Func<object, Task> _taskBody;
+            /// <summary>The <see cref="TaskScheduler"/> on which all work should be performed.</summary>
+            private readonly TaskScheduler _scheduler;
+            /// <summary>The <see cref="ExecutionContext"/> present at the time of the ForEachAsync invocation.  This is only used if on the default scheduler.</summary>
+            private readonly ExecutionContext? _executionContext;
+            /// <summary>Semaphore used to provide exclusive access to the enumerator.</summary>
+            private readonly SemaphoreSlim? _lock;
+
+            /// <summary>The number of outstanding workers.  When this hits 0, the operation has completed.</summary>
+            private int _completionRefCount;
+            /// <summary>Any exceptions incurred during execution.</summary>
+            private List<Exception>? _exceptions;
+            /// <summary>The number of workers that may still be created.</summary>
+            private int _remainingDop;
+
+            /// <summary>The delegate to invoke for each element yielded by the enumerator.</summary>
+            public readonly Func<TSource, CancellationToken, ValueTask> LoopBody;
+            /// <summary>The internal token source used to cancel pending work.</summary>
+            public readonly CancellationTokenSource Cancellation = new CancellationTokenSource();
+
+            /// <summary>Initializes the state object.</summary>
+            protected ForEachAsyncState(Func<object, Task> taskBody, bool needsLock, int dop, TaskScheduler scheduler, CancellationToken cancellationToken, Func<TSource, CancellationToken, ValueTask> body)
+            {
+                _taskBody = taskBody;
+                _lock = needsLock ? new SemaphoreSlim(initialCount: 1, maxCount: 1) : null;
+                _remainingDop = dop < 0 ? DefaultDegreeOfParallelism : dop;
+                LoopBody = body;
+                _scheduler = scheduler;
+                if (scheduler == TaskScheduler.Default)
+                {
+                    _executionContext = ExecutionContext.Capture();
+                }
+
+                _externalCancellationToken = cancellationToken;
+#if false
+                _registration = cancellationToken.UnsafeRegister(static o => ((ForEachAsyncState<TSource>)o!).Cancellation.Cancel(), this);
+#else
+                _registration = cancellationToken.Register(static o => ((ForEachAsyncState<TSource>)o!).Cancellation.Cancel(), this);
```

Comment:
i think this should be safe.

Reply:
Yes, the use of Unsafe is just an optimization to avoid unnecessarily capturing the current ExecutionContext (which is unfortunately more expensive on netfx).

Reply:
sgtm.